src/app/clusters/basic/basic.cpp

```diff
@@ -318,6 +318,101 @@ void emberAfBasicClusterServerInitCallback(chip::EndpointId endpoint)
     }
 }
 
+EmberAfStatus BasicClusterServerGetVendorIdAttribute(chip::EndpointId endpointId, chip::VendorId & value)
+{
+    return Attributes::VendorID::Get(endpointId, &value);
+}
+
+EmberAfStatus BasicClusterServerSetVendorIdAttribute(chip::EndpointId endpointId, chip::VendorId value)
```

Comment:
This should not be settable here: the spec has this Attribute as read-only. Let the platform decide how to set it, and return it properly from its `ConfigurationManager` interface implementation. The value is usually *fixed*. Same for all other Basic Information cluster attributes.

This setter is a layer-crossing that forces the value to always be RAM in the attribute store. 

Right now, the round trip is:

- Init product
- Call `emberAfBasicClusterServerInitCallback` in basic.cpp from init of data model
- Each attribute is set in the attribute store ram *after having being read from a constant API!*. This is a RAM cost that we need to remove.

The API added here prevents future refactor by enforcing that we ALWAYS have to round trip all these values from attribute store.

Reply:
That third item is basically about https://github.com/project-chip/connectedhomeip/issues/5926, right?

There are fundamentally two approaches we can take here.  Option 1 is that the value is pushed to the basic information cluster impl by other parts of the stack and it stores it however it wants to (doesn't have to be in RAM; could be configuration manager or whatnot, with AttributeAccessInterface to pull from there on read).  That's the option this API represents.  It does NOT force the value to be in RAM; it's just doing it for now because that's what we do right now anyway.  Later under the hood it can change.  Including to being a no-op (and then removed) if basic information is changed to do something more like option 2.

Option 2 is that the basic information cluster pulls the value from other parts of the stack via some API we define.  That could be the default value in the ZAP config (plus a possible fix for #5926), or it could be a delegate interface of some sort (including configuration manager queries under delegate here) plus an AttributeAccessInterface-based implementation of Basic Information.

It sounds like you are saying we should be doing option 2 to start with, right now, instead of taking the smaller step of just abstracting away the attribute access to make it easier to evolve to option 2 later if we want to?  Because that's what this PR is doing: taking that smaller step.

Reply:
Option 2 is nearly ready today.

Furthermore, the NotifyUpdateApplied does NOT need to look-up the value in the *cluster*. There is already a call available in ConfigurationManager to get current version.

To fix the issue of NotifyUpdateApplied, I would say it's easier to avoid any of the changes here for now, and for the following:

1. Have `NotifyUpdateApplied` API keep taking `version` and also add product ID
2. Let the *calling application* that knows, at some point, that NotifyUpdateApplied should be called, decide where/how to get the values
3. The example version today can take the value from ConfigurationManager directly. The `basic.cpp` file *already* inits all attribute values in attribute store from ConfigurationManager. The OTARequestor cluster does not need to get the value *from* the Basic Info cluster, it needs to have *the same value* as the Basic Info cluster. For our SDK, that value comes from ConfigurationManager, by platform abstraction API contract.

Reply:
The problem with that approach is that right now the basic information cluster only gets the value from the configuration manager on boot, right?  So are you arguing that we should couple the OTA changes to fixing that?

Reply:
The issue to be solved is calling `NotifyUpdateApplied` at the right time, with the right value.

The value of version can be obtained by the caller without tying OTARequestor impl to *any* BasicInfo cluster implementation.

In other words, we don't need to fix BasicInformationCluster for OTA. We just have to decouple OTARequestor from trying to get its data from BasicInformationCluster.

The spec says:

> After completion of an update, an OTA Requestor SHOULD invoke a `NotifyUpdateApplied` command to the OTA Provider which provided the initial query response to indicate that the OTA Requestor has successfully applied the OTA Software Update Image.

^^^ SHOULD, so right now it's not a blocker, since it's not a SHALL.

> [[OTA-NotifyUpdateApplied-SoftwareVersion]]

> ==== SoftwareVersion field
> The SoftwareVersion included in the request payload SHALL provide the same value as the <<ref_BasicInfoCluster_SoftwareVersion,`SoftwareVersion`>> attribute in the invoking OTA Requestor's <<ref_BasicInformationCluster,Basic Information Cluster>>, and SHOULD be consistent with the value representing a new version running on the Node invoking the command.
> 

^^^ The version has to be the same as what the BasicInformation cluster would report, but you don't need to *read it* from basic information cluster.